Add day selection screen - part 1

planarvoid opened the pull request:
Fixes #14705 

This PR adds functionality to select days for blogging reminders. This is only the first part of the UI changes because the code changes started to grow too much. The functionality to select a day for a reminder is there but it doesn't trigger a notification yet. 

What's missing
- polish the UI - especially fix the Diff utils for the day buttons so they don't all get refreshed when one is selected and the margins. Let's leave the design review for the second part.
- add the missing views under the Day buttons block

To test:
- Make sure the Blogging reminders flag is turned on
- Go to My Site/Settings
- Click on Blogging goals
- Notice the new updated UX
- Click on a few days 
- Confirm selection
- Notice the summary of the settings item is updated
- Click on a few Blogging goals
- Close bottom sheet
- Notice the changes are discarded and the summary is not updated


To test: 
- Turn on the app with "US English" language
- Notice the Days start with Sunday
- Change locale to some other language that starts week with Monday (for example Czech)
- Notice the Days start with Monday
- Check the same days as before are selected - only the days are shifted

<img width="501" alt="Screenshot 2021-06-02 at 18 11 13" src="https://user-images.githubusercontent.com/1079756/120515337-819a8980-c3ce-11eb-9c24-4c81d3e5f219.png">


## Regression Notes
1. Potential unintended areas of impact
None

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

planarvoid committed:
Add Top part of day selection screen

planarvoid committed:
Fix style issue

planarvoid committed:
Rename layout

planarvoid committed:
Fix test

planarvoid committed:
Fix view holder

planarvoid committed:
Add day buttons item to day selection screen

planarvoid committed:
Fix illustration drawing in Blogging reminders

planarvoid committed:
Make day bubbles in all caps

planarvoid committed:
Fix colors of the day buttons text

planarvoid committed:
Merge branch 'feature/14752-add-screens-switching-to-blogging-reminders' into feature/14705-add-day-selection-screen

planarvoid committed:
Fix ktlint issues

planarvoid committed:
Fix test

planarvoid committed:
Merge branch 'feature/14752-add-screens-switching-to-blogging-reminders' into feature/14705-add-day-selection-screen

planarvoid committed:
Merge branch 'feature/14752-add-screens-switching-to-blogging-reminders' into feature/14705-add-day-selection-screen

planarvoid committed:
Fix detekt issues

renanferrari commented on WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/DaysProvider.kt:
```diff
@@ -0,0 +1,33 @@
+class DaysProvider
+@Inject constructor(private val dateProvider: DateProvider) {
+    private val days = listOf(
+            Calendar.SUNDAY to SUNDAY,
+            Calendar.MONDAY to MONDAY,
+            Calendar.TUESDAY to TUESDAY,
+            Calendar.WEDNESDAY to WEDNESDAY,
+            Calendar.THURSDAY to THURSDAY,
+            Calendar.FRIDAY to FRIDAY,
+            Calendar.SATURDAY to SATURDAY
```
Even though this works pretty well (as the logic inside `DateProvider` does too), we should probably avoid using `Calendar` around this feature once #14698 gets merged and use `DayOfWeek` instead, since that's already being used in some places.

I know those parts of the code don't interact with each other right now, but I'm just trying to be on the careful side, since those two APIs have some significant differences when it comes to days of week (`Calendar.SUNDAY == 1` and `DayOfWeek.SUNDAY == 7` for example).


If we decide to start using desugaring within FluxC as well, then I guess this would actually be a requirement.

For now, I'd just add a `// TODO` on the top of this block to remind us of doing that. Let me know what you think!

planarvoid replied:
that's a good point. I agree 👍 . We can probably add a TODO and create a task to rewrite this once that PR is merged 👍

renanferrari commented on WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/DaysProvider.kt:
```diff
@@ -0,0 +1,33 @@
+            Calendar.SATURDAY to SATURDAY
+    )
+
+    fun getDays(): List<Pair<String, Day>> {
+        val offset = dateProvider.getFirstDayOfTheWeek() - 1
+        val shortWeekdays = dateProvider.getShortWeekdays()
+        val orderedDays = days.takeLast(days.size - offset) + days.take(offset)
+        return orderedDays.map { shortWeekdays[it.first] to it.second }
+    }
```
Also, if we start using `DayOfWeek` within our `BloggingRemindersModel` I think this logic could be simplified quite a bit (maybe even removed entirely).

renanferrari reviewed (commented):
Thanks for the changes! 

The app is working as expected for the most part, except that I see these weird animations when opening/closing the bottom sheet:

https://user-images.githubusercontent.com/830056/120874909-307dc780-c57f-11eb-9fa8-4830964843cd.mp4

Are you planning to tackle that as part of the UI polish in the second part?

As for the code, it is looking good overall, I just have a few observations which I've shared below.

planarvoid committed:
Add TODO to DaysProvider.kt

planarvoid committed:
Remove unused resource

planarvoid commented:
Thanks for the review @renanferrari . I've added the TODO and fixed a lint issue. I was wondering about that UI behaviour. I couldn't reproduce it either my emulator nor my pixel. Do you have repro steps? Or does it always just happen to you?

renanferrari reviewed (approved):
Thanks for the changes!

>I was wondering about that UI behaviour. I couldn't reproduce it either my emulator nor my pixel.

I actually tried this again today and couldn't reproduce it myself (using the same device and everything), so I don't know what that was about 🤷‍♂️

I'm gonna merge this and if we come across that bug again we can open a separate issue for it.
